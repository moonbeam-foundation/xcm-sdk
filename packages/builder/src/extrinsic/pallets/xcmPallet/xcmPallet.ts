import { ExtrinsicConfig } from '../../../types/substrate/ExtrinsicConfig';
import {
  type ExtrinsicConfigBuilder,
  type Parents,
  XcmVersion,
} from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicAccount,
  getExtrinsicArgumentVersion,
  normalizeConcrete,
} from '../../ExtrinsicBuilder.utils';
import { getPolkadotXcmExtrinsicArgs } from '../polkadotXcm/polkadotXcm.util';

const pallet = 'xcmPallet';

export function xcmPallet() {
  return {
    limitedReserveTransferAssets: (parents: Parents = 1) => {
      const func = 'limitedReserveTransferAssets';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);
                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  parents,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(version, {
                        parents: 0,
                        interior: 'Here',
                      }),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                  ],
                });
              },
            }),
        }),
      };
    },
    transferAssetsUsingTypeAndThen: () => {
      const func = 'transferAssetsUsingTypeAndThen';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: () => {
                const version = XcmVersion.v4;
                return [
                  {
                    [version]: {
                      parents: 0,
                      interior: {
                        X1: [
                          {
                            Parachain: params.destination.parachainId,
                          },
                        ],
                      },
                    },
                  },
                  {
                    [version]: [
                      {
                        id: {
                          parents: 0,
                          interior: 'Here',
                        },
                        fun: {
                          Fungible: params.asset.amount,
                        },
                      },
                    ],
                  },
                  'LocalReserve',
                  {
                    [version]: {
                      parents: 0,
                      interior: 'Here',
                    },
                  },
                  'LocalReserve',
                  {
                    [version]: [
                      {
                        DepositAsset: {
                          assets: {
                            Wild: { AllCounted: 1 },
                          },
                          beneficiary: {
                            parents: 0,
                            interior: {
                              X1: [
                                getExtrinsicAccount(params.destinationAddress),
                              ],
                            },
                          },
                        },
                      },
                    ],
                  },
                  'Unlimited',
                ];
              },
            }),
        }),
      };
    },
  };
}
