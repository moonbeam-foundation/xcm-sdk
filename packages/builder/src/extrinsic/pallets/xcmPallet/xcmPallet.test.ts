import { describe, expect, it } from 'vitest';

import { buildParamsMock } from '../../../../fixtures';
import { xcmPallet } from './xcmPallet';

describe('xcmPallet', () => {
  describe('limitedReserveTransferAssets', () => {
    const extrinsic = xcmPallet()
      .limitedReserveTransferAssets()
      .here()
      .build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs()).toMatchSnapshot();
    });
  });
});
