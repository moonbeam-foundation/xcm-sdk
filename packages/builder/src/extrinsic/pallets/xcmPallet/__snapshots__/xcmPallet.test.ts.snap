// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`xcmPallet > limitedReserveTransferAssets > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "xcmPallet",
}
`;

exports[`xcmPallet > limitedReserveTransferAssets > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;
